Retry if RPM lock is temporarily unavailable

witekest opened the pull request:
### What does this PR do?

Wait and retry if Zypper fails acquiring RPM lock file.

### What issues does this PR fix or reference?
Fixes: SUSE/spacewalk#18163

### Previous Behavior
Reportedly highstate was failing in case when RPM lock file was used by some other process.

### New Behavior
Wait and retry if Zypper fails acquiring RPM lock file.

### Merge requirements satisfied?
- [ ] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
No

witekest committed:
Retry if RPM lock is temporarily unavailable

witekest committed:
Add changelog entry

meaksh commented on salt/modules/zypperpkg.py:
```diff
@@ -240,6 +242,12 @@ def _is_lock(self):
+        """
+        Is this an RPM lock error?
+        """
+        return self.RPM_LOCK_ERR_MESSAGE in self.error_msg
```
We need to be more strict here to consider the error is actually an RPM lock error.

If we only check if that message exists as part of the error coming from Zypper, we can get the minion into and infitine loop when passing that string as a malicious package name, like:

```
# salt-call pkg.install "can't create transaction lock on"
```

vzhestkov replied:
With the other locale on the system the message could be different

witekest replied:
The problem is the exit code from RPM does not specify what kind of error has actually occurred. Exit code is equal to the number of errors of any kind which have happened during the execution.
What we could do here is to simulate RPM behavior and try to acquire the file lock. Would that be the preferred approach?

meaksh reviewed (changes requested):
Thanks for this PR @witekest ! :+1:

I did some comments spotting some things I think we should improve.

witekest committed:
Check if RPM file lock can get acquired

witekest committed:
Add complete err object to data[]

Co-authored-by: Pablo Suárez Hernández <psuarezhernandez@suse.com>

witekest committed:
Simple handling of RPM lock file

Differently as Zypper RPM does not write the process ID into the lock
file. Therefore RPM file lock has to be handled seperately.

witekest commented:
I've split now the handling of Zypper and RPM locks as the latter does not store information about the locking process. I've tested it by manually acquiring exclusive lock over the RPM lock file. Works as expected and retries the installation until the lock gets released.

witekest committed:
Fix existing unit tests

witekest committed:
Merge branch 'master' into rpm_lock

meaksh commented on tests/unit/modules/test_zypperpkg.py:
```diff
@@ -94,7 +94,7 @@ def test_list_upgrades(self):
         with patch.dict(
             zypper.__salt__, {"cmd.run_all": MagicMock(return_value=ref_out)}
-        ):
+        ), patch.object(zypper.__zypper__, "_is_rpm_lock", return_value=False):
```
I guess this should be like:
```suggestion
        ), patch.object(zypper.__zypper__, "_is_rpm_lock", MagicMock(return_value=False)):
```

as `return_value` is a parameter for `MagicMock`, right?

meaksh replied:
Good to know, thanks for clarifying! :+1:

meaksh commented on tests/unit/modules/test_zypperpkg.py:
```diff
@@ -194,7 +194,8 @@ def __call__(self, *args, **kwargs):
         sniffer = RunSniffer(stdout=stdout_xml_snippet, retcode=1)
-        with patch.dict("salt.modules.zypperpkg.__salt__", {"cmd.run_all": sniffer}):
+        with patch.dict("salt.modules.zypperpkg.__salt__", {"cmd.run_all": sniffer}), \
+            patch.object(zypper.__zypper__, "_is_rpm_lock", return_value=False):
```
Same here:
```suggestion
            patch.object(zypper.__zypper__, "_is_rpm_lock", MagicMock(return_value=False)):
```

meaksh commented on tests/unit/modules/test_zypperpkg.py:
```diff
@@ -241,7 +242,7 @@ def test_list_upgrades_error_handling(self):
             "salt.modules.zypperpkg.__salt__",
             {"cmd.run_all": MagicMock(return_value=ref_out)},
-        ):
+        ), patch.object(zypper.__zypper__, "_is_rpm_lock", return_value=False):
```
```suggestion
        ), patch.object(zypper.__zypper__, "_is_rpm_lock", MagicMock(return_value=False)):
```

meaksh commented on tests/unit/modules/test_zypperpkg.py:
```diff
@@ -228,7 +229,7 @@ def test_list_upgrades_error_handling(self):
             "salt.modules.zypperpkg.__salt__",
             {"cmd.run_all": MagicMock(return_value=ref_out)},
-        ):
+        ), patch.object(zypper.__zypper__, "_is_rpm_lock", return_value=False):
```
Same here:
```suggestion
        ), patch.object(zypper.__zypper__, "_is_rpm_lock", MagicMock(return_value=False)):
```

meaksh commented:
@witekest it seems these changes causes some failures in "windows" as `fcntl` is not available.

We should try to avoid this, for example by doing "try/except" on the import as it is done here: https://github.com/saltstack/salt/blob/master/salt/utils/files.py#L26-L32

witekest committed:
Do not acquire lock on systems without fcntl

witekest commented:
> @witekest it seems these changes causes some failures in "windows" as `fcntl` is not available.
> 
> We should try to avoid this, for example by doing "try/except" on the import as it is done here: https://github.com/saltstack/salt/blob/master/salt/utils/files.py#L26-L32

Thanks. Good hint. Updated the code accordingly.

witekest committed:
Add unit tests for _is_rpm_lock

Co-authored-by: Pablo Suárez Hernández <psuarezhernandez@suse.com>

witekest committed:
Merge branch 'master' into rpm_lock

witekest committed:
Do not run fcntl tests on Windows

witekest committed:
Fix fcntl unit test

witekest committed:
Rerun checks

witekest committed:
Merge branch 'master' into rpm_lock

witekest committed:
Merge branch 'master' into rpm_lock

witekest committed:
Use legacy file lock location

SLE12 systems use /var/lib/rpm/.rpm.lock as file lock. SLE15 is
backwards compatible by using symbolic link.

meaksh reviewed (dismissed):
Looks good! Thanks! :+1:

witekest committed:
Re-trigger CI checks

witekest committed:
Merge branch 'master' into rpm_lock

witekest committed:
Merge branch 'master' into rpm_lock

witekest commented:
The users have reportedly validated the code by testing it in their environment. Feedback was positive.

witekest committed:
Merge branch 'master' into rpm_lock

nicholasmhughes commented:
re-run pr-ubuntu-2204-amd64-py3-m2crypto-pytest

nicholasmhughes commented:
@witekest it looks like pre-commit is failing isort and black checks. Please check the [Contributing documentation](https://docs.saltproject.io/en/latest/topics/development/contributing.html#pre-commit-and-nox-setup) for information on installing and using pre-commit. Thank you!

witekest committed:
Fix linter errors

witekest committed:
Retry running pre-commit

witekest commented:
Thanks!
